File: tests/math_cpp/composite_bound.cpp
// Composite SVS security bound — C++ port of tests/math/composite_bound.py.
// Sessions 1-6 sharpenings folded in (see Python source for full commentary).
//
// Self-contained: no external deps beyond libc++.

#include <cmath>
#include <cstdio>
#include <string>

namespace {

struct Result {
    double T1, T2, T3, T4;
    double T5_kb;
    double T_crypto, total;
    int m, N, k, Q, num_sectors;
    double eps_agg, eps_card, eps_cell;
    std::string composition;
    bool include_dp;
};

double T1_crypto(int m) {
    double lg = std::log2(m);
    double H_inf = m * lg / 2.0;
    return std::pow(2.0, -H_inf);
}

double T2_incidence(int k, int /*N*/, double eps_cell) {
    double padding_per_cell = 1.0 / (std::exp(eps_cell) - 1.0);
    return 1.0 / (k + padding_per_cell);
}

double T3_dp_composition(int Q, int num_sectors,
                         double eps_agg, double eps_card,
                         const std::string& composition) {
    double per_query = eps_agg + eps_card;
    double total_queries = static_cast<double>(Q) * num_sectors;
    if (composition == "basic") {
        return total_queries * per_query;
    } else if (composition == "renyi_alpha2") {
        // δ = 2^-40; ln(1/δ) ≈ 27.7
        return std::sqrt(2.0 * total_queries * 27.7) * per_query
             + total_queries * per_query * per_query;
    }
    return -1.0;
}

double T4_single_benes_dist(int m, double friedrichs_cos = 0.5) {
    int D = 2 * static_cast<int>(std::log2(m)) - 1;
    return std::pow(friedrichs_cos, D - 1);
}

double T5_comm_cost_kbytes(int m) {
    int D = 2 * static_cast<int>(std::log2(m)) - 1;
    int osn_calls = 2 * D + 1;
    int kappa = 128;
    double per_osn_bits = 2.0 * kappa * m * static_cast<int>(std::log2(m));
    return osn_calls * per_osn_bits / 8.0 / 1024.0;
}

Result composite_bound(int m, int N, int k, int Q, int num_sectors,
                       double eps_agg, double eps_card, double eps_cell,
                       const std::string& composition = "basic",
                       bool include_dp = true) {
    Result r{};
    r.T1 = T1_crypto(m);
    r.T2 = T2_incidence(k, N, eps_cell);
    r.T3 = include_dp
        ? T3_dp_composition(Q, num_sectors, eps_agg, eps_card, composition)
        : 0.0;
    r.T4 = T4_single_benes_dist(m);
    r.T5_kb = T5_comm_cost_kbytes(m);
    r.T_crypto = std::max(r.T1, r.T4);
    r.total = r.T_crypto + r.T2 + r.T3;
    r.m = m; r.N = N; r.k = k; r.Q = Q; r.num_sectors = num_sectors;
    r.eps_agg = eps_agg; r.eps_card = eps_card; r.eps_cell = eps_cell;
    r.composition = composition; r.include_dp = include_dp;
    return r;
}

std::string fmt_sci(double x, int sig = 3) {
    char buf[64];
    if (x == 0.0)     { std::snprintf(buf, sizeof(buf), "0");            return buf; }
    if (x < 1e-30)    { std::snprintf(buf, sizeof(buf), "< 2^-100");     return buf; }
    if (x < 0.001)    { std::snprintf(buf, sizeof(buf), "%.*e", sig, x); return buf; }
    std::snprintf(buf, sizeof(buf), "%.*f", sig, x);
    return buf;
}

void print_scenario(const std::string& label, const Result& r) {
    std::printf("--- %s ---\n", label.c_str());
    std::printf("  Params: m=%d, N=%d, k=%d, Q=%d, sectors=%d, "
                "eps_agg=%g, eps_card=%g, eps_cell=%g, composition=%s, DP=%s\n",
                r.m, r.N, r.k, r.Q, r.num_sectors,
                r.eps_agg, r.eps_card, r.eps_cell,
                r.composition.c_str(), r.include_dp ? "true" : "false");
    std::printf("  T1 (crypto guess):     %s\n", fmt_sci(r.T1).c_str());
    std::printf("  T4 (single-Benes TV):  %s   [session 6]\n", fmt_sci(r.T4).c_str());
    std::printf("  T_crypto = max(T1,T4): %s\n", fmt_sci(r.T_crypto).c_str());
    std::printf("  T2 (incidence):        %s\n", fmt_sci(r.T2).c_str());
    std::printf("  T3 (DP output):        %s%s\n", fmt_sci(r.T3).c_str(),
                r.include_dp ? "" : "   [DISABLED]");
    std::printf("  Total advantage:       %s\n", fmt_sci(r.total).c_str());
    std::printf("  T5 comms per query:    %.0f KB   [session 5, not Adv]\n", r.T5_kb);

    struct Term { const char* name; double val; };
    Term ts[] = { {"T_crypto", r.T_crypto}, {"T2", r.T2}, {"T3", r.T3} };
    Term* binding = &ts[0];
    for (int i = 1; i < 3; ++i) if (ts[i].val > binding->val) binding = &ts[i];
    std::printf("  BINDING TERM: %s  (%s)\n\n",
                binding->name, fmt_sci(binding->val).c_str());
}

}  // namespace

int main() {
    std::puts("======================================================================");
    std::puts("Composite SVS security bound - numerical evaluator (C++ port)");
    std::puts("(Sessions 1-6 sharpenings folded in.)");
    std::puts("======================================================================");
    std::puts("");
    std::puts("Coalition scope for T_crypto: post-sessions-4/5 (both-blind + two-server)");
    std::puts("  -> covers 5 non-trivial coalitions (vs 1 in single-server)");
    std::puts("  -> same gamma(D) bound, ~5x broader threat model");
    std::puts("");
    std::printf("Session 6 addition:\n");
    std::printf("  T4_dist(4)     = %s  (verified concretely, cos theta' = 0.5)\n",
                fmt_sci(T4_single_benes_dist(4)).c_str());
    std::printf("  T4_dist(1024)  = %s  (optimistic conjecture)\n",
                fmt_sci(T4_single_benes_dist(1024)).c_str());
    std::printf("  T4_dist(16384) = %s  (optimistic conjecture)\n\n",
                fmt_sci(T4_single_benes_dist(16384)).c_str());
    std::printf("Session 5 comms cost:\n");
    std::printf("  T5 at m=1024:  %.0f KB per query\n", T5_comm_cost_kbytes(1024));
    std::printf("  T5 at m=16384: %.0f KB per query\n\n", T5_comm_cost_kbytes(16384));

    print_scenario("A. MAS default (basic DP composition)",
                   composite_bound(1024, 4, 25, 4, 20, 0.5, 0.1, 1.0, "basic"));
    print_scenario("B. MAS default (Renyi alpha=2 composition)",
                   composite_bound(1024, 4, 25, 4, 20, 0.5, 0.1, 1.0, "renyi_alpha2"));
    print_scenario("C. Reduced granularity: 5 sectors",
                   composite_bound(1024, 4, 25, 4, 5, 0.5, 0.1, 1.0, "basic"));
    print_scenario("D. Small per-query eps (0.01 each)",
                   composite_bound(1024, 4, 25, 4, 20, 0.01, 0.01, 1.0, "basic"));
    print_scenario("E. Annual release only (Q=1)",
                   composite_bound(1024, 4, 25, 1, 20, 0.5, 0.1, 1.0, "basic"));
    print_scenario("F. Sweet spot: annual, 5 sectors, small eps",
                   composite_bound(1024, 4, 25, 1, 5, 0.1, 0.05, 1.0, "basic"));
    print_scenario("G. Large-population (m=16K, N=6)",
                   composite_bound(16384, 6, 50, 4, 20, 0.5, 0.1, 1.0, "basic"));
    print_scenario("H. Sub-unity sweet spot: Q=1, 5 sectors, eps=0.02",
                   composite_bound(1024, 4, 25, 1, 5, 0.02, 0.01, 1.0, "basic"));
    print_scenario("I. Single-scalar release, semi-annual",
                   composite_bound(1024, 4, 25, 2, 1, 0.05, 0.02, 1.0, "basic"));

    std::puts("======================================================================");
    std::puts("CURRENT DEPLOYMENT (DP disabled, both-blind + two-server enabled)");
    std::puts("======================================================================\n");
    print_scenario("J. Current deployment (MAS quarterly, DP off)",
                   composite_bound(1024, 4, 25, 4, 20, 0.0, 0.0, 1.0, "basic", false));
    print_scenario("K. Current deployment, large population (m=16K)",
                   composite_bound(16384, 6, 50, 4, 20, 0.0, 0.0, 1.0, "basic", false));
    print_scenario("L. Current deployment, k=1 (no k-anon padding)",
                   composite_bound(1024, 4, 1, 4, 20, 0.0, 0.0, 1.0, "basic", false));

    std::puts("======================================================================");
    std::puts("KEY FINDINGS (see composite_bound.py for full narrative)");
    std::puts("======================================================================");
    std::puts("Regime 2 (DP off) headline numbers:");
    std::puts("  J. Adv = 0.039 at MAS default (k=25)");
    std::puts("  K. Adv = 0.020 at m=16K, k=50");
    std::puts("  L. Adv = 0.632 at k=1 (k-anon is load-bearing)");
    std::puts("");
    std::puts("Session 4/5/6 coalition/comms/mixing overlays as summarized in .py.");
    return 0;
}

File: tests/math/composite_bound.py
r"""
Composite SVS security bound — numerical evaluator.

Implements the theorem from docs/COMPOSITE_SECURITY_THEOREM.md as a
callable function. Sweeps operating parameters and prints a table showing
where each term dominates.

Session-by-session refinements folded in:

  Session 1 (sharp H_∞):  T1_crypto uses m·log₂m/2, not m log m − Θ(m).
  Session 2 (γ(D) tight): confirms T1_crypto is the correct upper bound.
  Session 3 (this file's original scope): identified T3 as dominant.
  Session 4 (both-blind XOR variant): T1's coalition scope expands —
     Matcher+Party can no longer recover ρ_2. Same numerical T1, broader
     coverage (see coalition_scope() below).
  Session 5 (two-server ρ_2 protocol): adds 2D+1 OSN comms cost per query
     but zero Adv cost. Reported separately as T5_comm.
  Session 6 (spectral/Friedrichs): adds T4_dist — single-Beneš TV bound
     from the mixing-rate analysis. Complements T1 (guessing) with a
     distribution-distinguishing bound.

Current operational stance (per user direction 2026-Q3):
  - DP output layer DEPRIORITIZED: T3 kept for reference but not the
    binding constraint. Adv = T1 + T2 + T4 in current deployment.
  - Both-blind + two-server protocol assumed deployed (T1 coalition scope
    is the "sessions 4/5" version).
"""

import math


def T1_crypto(m):
    r"""T1 = γ(D) upper bound, using session-1's sharp form H_∞ = m log₂m / 2.
    Returns Adv upper bound as a positive real; ~ 0 for realistic m.

    Coverage (post-sessions 4/5, both-blind + two-server protocol assumed):
      - Adversary = any single party (matcher OR router OR sender)
      - Adversary = matcher+party OR router+party coalition
      - Excluded: matcher+router (non-collusion assumption)
    """
    lg = math.log2(m)
    H_inf = m * lg / 2
    return 2 ** (-H_inf)


def T4_single_benes_dist(m, friedrichs_cos=0.5):
    r"""
    Session 6 finding: for k=1 Beneš (as deployed), residual TV(D, Unif)
    is bounded by the operator norm of D̂ on the mixing-bottleneck irrep.

    By von Neumann product-of-projections: ||D̂||_op ≤ (cos θ')^(D-1)
    where θ' is the Friedrichs angle between post-wiring column stabilisers.
    At m=4 we verified cos θ' = 0.5 exactly; conjecture it's ≤ 0.5 in general
    (open for m > 4).

    Optimistic bound (cos θ' = 0.5 across all m):
      ||D̂||_op ≤ 0.5^(D-1) = 0.5^(2·log₂m − 2) = 4/m²
      At m=1024: ~3.8e-6.

    This is a DIFFERENT metric from T1: T1 is the guessing probability, T4
    is the distinguishing advantage. Take the max in a security union bound
    if the adversary can leverage either.

    NOTE: for OIRA at m=1024, T1 (2^-5120) is astronomically smaller than
    T4 (4e-6). But T4 bounds a MEASURABLE statistical divergence (party
    could run a chi-square test on many outputs to detect the bias),
    whereas T1 bounds a one-shot guess. Both are needed for a full picture.
    """
    if not (m >= 2 and (m & (m - 1)) == 0):
        raise ValueError(f"m must be a power of 2, got {m}")
    D = 2 * int(math.log2(m)) - 1
    return friedrichs_cos ** (D - 1)


def T5_comm_cost_kbytes(m):
    r"""
    Session 5: per-party query cost of the two-server ρ_2 protocol.
    2D+1 OSN calls, ~2·κ·m·log₂m bits per OSN at κ=128.

    Not an Adv term. Reported for operational cost budgeting.
    """
    D = 2 * int(math.log2(m)) - 1
    osn_calls = 2 * D + 1
    kappa = 128
    per_osn_bits = 2 * kappa * m * int(math.log2(m))
    total_bytes = osn_calls * per_osn_bits / 8
    return total_bytes / 1024


def coalition_scope():
    r"""
    Human-readable summary of which coalitions T1's bound covers.

    Pre-session-4 (single-server variant, deployed):
      - Matcher alone: KNOWS π (no bound needed)
      - Router alone: N/A (no router)
      - Party alone: guessing prob ≤ γ(D)
      - Matcher+Party: matcher has π, party has ρ_2 — matcher already knows
        everything, no incremental coverage
      → Only 1 non-trivial coalition covered.

    Post-session-4/5 (both-blind + two-server, deployable):
      - Matcher alone: has (π, a), CANNOT compute ρ_1 or ρ_2 (needs b)
      - Router alone: has b only, no π/ρ_1/ρ_2
      - Party alone: has ρ_2, guessing prob ≤ γ(D)
      - Matcher+Party: has (π, a, ρ_2), CANNOT recover ρ_1 (missing b)
      - Router+Party: has (b, ρ_2), CANNOT recover π (missing a)
      - Matcher+Router (EXCLUDED, non-collusion): trivial full leak
      → 5 non-trivial coalitions covered by the same γ(D) bound.

    Same numerical bound, ~5× broader threat model — a strict security
    improvement at zero Adv cost, with ~1.5 MB comms overhead per query
    (T5_comm_cost_kbytes).
    """
    return "post-sessions-4/5 (both-blind + two-server)"


def T2_incidence(k, N, eps_cell):
    r"""T2 = 1 / (k + padding_floor) where padding_floor is Mechanism 5.3
    expected per-cell overhead added to noise floor."""
    padding_floor = (2 ** N - 1) / (math.exp(eps_cell) - 1) / (2 ** N - 1)
    # Actually padding per cell is (2^N - 1)/(e^ε - 1) TOTAL, divided by (2^N - 1) cells
    # = 1/(e^ε - 1) per cell on average.
    # Recomputed cleanly:
    padding_per_cell = 1.0 / (math.exp(eps_cell) - 1)
    return 1.0 / (k + padding_per_cell)


def T3_dp_composition(Q, num_sectors, eps_agg, eps_card, composition="basic"):
    r"""T3 = DP composition cost across Q sessions and num_sectors sectors."""
    per_query = eps_agg + eps_card
    total_queries = Q * num_sectors
    if composition == "basic":
        return total_queries * per_query
    elif composition == "renyi_alpha2":
        # Rough Rényi at α=2 bound: sqrt(2 * total_queries * ln(1/δ)) * eps + total_queries * eps^2
        # With δ = 2^-40, ln(1/δ) ≈ 27.7
        import math
        return math.sqrt(2 * total_queries * 27.7) * per_query + total_queries * per_query ** 2
    else:
        raise ValueError(f"unknown composition mode: {composition}")


def composite_bound(m, N, k, Q, num_sectors, eps_agg, eps_card, eps_cell,
                    composition="basic", include_dp=True):
    r"""
    Composite Adv bound. Returns dict with each term + total.

    Terms:
      T1 — crypto guessing bound γ(D) = 2^{-m log₂m/2}. Session 1 (sharp).
           Coalition scope: sessions 4/5 (both-blind + two-server).
      T2 — Tool 5 incidence: identification prob 1/(k + padding).
      T3 — DP output composition. Deprioritized 2026-Q3 (include_dp toggles).
      T4 — session 6 single-Beneš TV bound = 4/m². Complements T1.
      T5 — session 5 comms cost in KB. Not an Adv term; reported separately.
    """
    T1 = T1_crypto(m)
    T2 = T2_incidence(k, N, eps_cell)
    T3 = T3_dp_composition(Q, num_sectors, eps_agg, eps_card, composition) if include_dp else 0.0
    T4 = T4_single_benes_dist(m)
    T5_kb = T5_comm_cost_kbytes(m)
    # Adv is a union bound over adversary strategies.
    # T1 (guess π specifically) and T4 (distinguish D from Unif) are ALTERNATIVE
    # metrics for the same crypto tier — max, not sum. T2, T3 are additive.
    T_crypto = max(T1, T4)
    total = T_crypto + T2 + T3
    return {"T1": T1, "T2": T2, "T3": T3, "T4": T4, "T5_kb": T5_kb,
            "T_crypto": T_crypto, "total": total,
            "params": {"m": m, "N": N, "k": k, "Q": Q,
                       "num_sectors": num_sectors,
                       "eps_agg": eps_agg, "eps_card": eps_card,
                       "eps_cell": eps_cell,
                       "composition": composition,
                       "include_dp": include_dp}}


def fmt_sci(x, sig=3):
    if x == 0:
        return "0"
    if x < 1e-30:
        return f"< 2^-100"
    if x < 0.001:
        return f"{x:.{sig}e}"
    return f"{x:.{sig}f}"


def print_scenario(label, result):
    p = result["params"]
    print(f"--- {label} ---")
    print(f"  Params: m={p['m']}, N={p['N']}, k={p['k']}, Q={p['Q']}, "
          f"sectors={p['num_sectors']}, "
          f"ε_agg={p['eps_agg']}, ε_card={p['eps_card']}, ε_cell={p['eps_cell']}, "
          f"composition={p['composition']}, DP={p['include_dp']}")
    print(f"  T1 (crypto guess):     {fmt_sci(result['T1'])}")
    print(f"  T4 (single-Beneš TV):  {fmt_sci(result['T4'])}   [session 6]")
    print(f"  T_crypto = max(T1,T4): {fmt_sci(result['T_crypto'])}")
    print(f"  T2 (incidence):        {fmt_sci(result['T2'])}")
    print(f"  T3 (DP output):        {fmt_sci(result['T3'])}"
          + ("" if p['include_dp'] else "   [DISABLED]"))
    print(f"  Total advantage:       {fmt_sci(result['total'])}")
    print(f"  T5 comms per query:    {result['T5_kb']:.0f} KB   [session 5, not Adv]")
    # Identify binding constraint
    terms = [("T_crypto", result['T_crypto']),
             ("T2", result['T2']),
             ("T3", result['T3'])]
    binding = max(terms, key=lambda x: x[1])
    print(f"  BINDING TERM: {binding[0]}  ({fmt_sci(binding[1])})")
    print()


def main():
    print("=" * 70)
    print("Composite SVS security bound — numerical evaluator")
    print("(Sessions 1-6 sharpenings folded in.)")
    print("=" * 70)
    print()
    print(f"Coalition scope for T_crypto: {coalition_scope()}")
    print(f"  → covers 5 non-trivial coalitions (vs 1 in single-server)")
    print(f"  → same γ(D) bound, ~5× broader threat model")
    print()
    print("Session 6 addition:")
    print(f"  T4_dist(m) at m=4     = {fmt_sci(T4_single_benes_dist(4))}  "
          f"(verified concretely, cos θ' = 0.5)")
    print(f"  T4_dist(m) at m=1024  = {fmt_sci(T4_single_benes_dist(1024))}  "
          f"(optimistic conjecture)")
    print(f"  T4_dist(m) at m=16384 = {fmt_sci(T4_single_benes_dist(16384))}  "
          f"(optimistic conjecture)")
    print()
    print("Session 5 comms cost:")
    print(f"  T5 at m=1024:  {T5_comm_cost_kbytes(1024):.0f} KB per query")
    print(f"  T5 at m=16384: {T5_comm_cost_kbytes(16384):.0f} KB per query")
    print()
    print("Scenarios exercise different operating points to identify where")
    print("each term dominates.")
    print()

    # Scenario A: MAS default from PROBLEM_STATEMENT.md
    r = composite_bound(
        m=1024, N=4, k=25, Q=4, num_sectors=20,
        eps_agg=0.5, eps_card=0.1, eps_cell=1.0,
        composition="basic")
    print_scenario("A. MAS default (basic DP composition)", r)

    # Scenario B: Same, with Rényi DP composition
    r = composite_bound(
        m=1024, N=4, k=25, Q=4, num_sectors=20,
        eps_agg=0.5, eps_card=0.1, eps_cell=1.0,
        composition="renyi_alpha2")
    print_scenario("B. MAS default (Rényi α=2 composition)", r)

    # Scenario C: Reduced granularity — 5 sectors instead of 20
    r = composite_bound(
        m=1024, N=4, k=25, Q=4, num_sectors=5,
        eps_agg=0.5, eps_card=0.1, eps_cell=1.0,
        composition="basic")
    print_scenario("C. Reduced granularity: 5 sectors", r)

    # Scenario D: Very small per-query ε
    r = composite_bound(
        m=1024, N=4, k=25, Q=4, num_sectors=20,
        eps_agg=0.01, eps_card=0.01, eps_cell=1.0,
        composition="basic")
    print_scenario("D. Small per-query ε (0.01 each)", r)

    # Scenario E: Annual release (Q=1) — no accumulation across quarters
    r = composite_bound(
        m=1024, N=4, k=25, Q=1, num_sectors=20,
        eps_agg=0.5, eps_card=0.1, eps_cell=1.0,
        composition="basic")
    print_scenario("E. Annual release only (Q=1)", r)

    # Scenario F: Sweet spot — reduced sectors + small ε + annual
    r = composite_bound(
        m=1024, N=4, k=25, Q=1, num_sectors=5,
        eps_agg=0.1, eps_card=0.05, eps_cell=1.0,
        composition="basic")
    print_scenario("F. Sweet spot: annual, 5 sectors, small ε", r)

    # Scenario G: Large-population MAS full-population run
    r = composite_bound(
        m=16384, N=6, k=50, Q=4, num_sectors=20,
        eps_agg=0.5, eps_card=0.1, eps_cell=1.0,
        composition="basic")
    print_scenario("G. Large-population (m=16K, N=6)", r)

    # Scenario H: True sweet spot — annual, tiny ε
    r = composite_bound(
        m=1024, N=4, k=25, Q=1, num_sectors=5,
        eps_agg=0.02, eps_card=0.01, eps_cell=1.0,
        composition="basic")
    print_scenario("H. Sub-unity sweet spot: Q=1, 5 sectors, ε=0.02", r)

    # Scenario I: Aggressive sub-unity — semi-annual, single global agg
    r = composite_bound(
        m=1024, N=4, k=25, Q=2, num_sectors=1,
        eps_agg=0.05, eps_card=0.02, eps_cell=1.0,
        composition="basic")
    print_scenario("I. Single-scalar release, semi-annual", r)

    # -- Current-deployment scenarios (2026-Q3): DP disabled per user direction --
    print("=" * 70)
    print("CURRENT DEPLOYMENT (DP disabled, both-blind + two-server enabled)")
    print("=" * 70)
    print()

    # Scenario J: Current deployment, MAS quarterly full sweep, DP off
    r = composite_bound(
        m=1024, N=4, k=25, Q=4, num_sectors=20,
        eps_agg=0.0, eps_card=0.0, eps_cell=1.0,
        composition="basic", include_dp=False)
    print_scenario("J. Current deployment (MAS quarterly, DP off)", r)

    # Scenario K: Current deployment, larger population
    r = composite_bound(
        m=16384, N=6, k=50, Q=4, num_sectors=20,
        eps_agg=0.0, eps_card=0.0, eps_cell=1.0,
        composition="basic", include_dp=False)
    print_scenario("K. Current deployment, large population (m=16K)", r)

    # Scenario L: Current deployment, k=1 (smallest padding)
    r = composite_bound(
        m=1024, N=4, k=1, Q=4, num_sectors=20,
        eps_agg=0.0, eps_card=0.0, eps_cell=1.0,
        composition="basic", include_dp=False)
    print_scenario("L. Current deployment, k=1 (no k-anon padding)", r)

    print("=" * 70)
    print("KEY FINDINGS (with sessions 1-6 sharpenings)")
    print("=" * 70)
    print()
    print("REGIME 1 — Historical (with DP output layer, session 3 finding):")
    print()
    print("1. T_crypto = max(T1, T4). At m=1024, T1 ~ 2^-5120 (session 1)")
    print("   and T4 ~ 4e-6 (session 6). T4 DOMINATES T1 by 5000 orders of")
    print("   magnitude, but is still tiny — crypto tier not binding.")
    print()
    print("2. T2 (incidence) is ~0.03 with k=25 — the leakage floor.")
    print()
    print("3. T3 (DP composition) is DOMINANT under basic composition:")
    print("   scenarios A, B (Q=4, 20 sectors) give Adv > 1. Sub-unity")
    print("   requires annual cadence + ≤5 sectors + ε ≤ 0.02 (H, I).")
    print()
    print("REGIME 2 — Current deployment (DP disabled per 2026-Q3 direction):")
    print()
    print("4. With T3 = 0, the binding term is T2 (incidence padding floor).")
    print("   Scenario J: Adv ≈ 0.038 — 26× tighter than uniform baseline.")
    print("   Scenario K (m=16K): T4 ≈ 4/16K² ~ 1.5e-8, T2 unchanged.")
    print()
    print("5. Session 4 broadens the coalition scope of T_crypto without")
    print("   changing its numerical value: matcher+party can no longer")
    print("   recover ρ_2 (needs router's b share). 5× more coalitions")
    print("   covered by the same bound.")
    print()
    print("6. Session 5 protocol cost: ~12 MB comms per party query at")
    print("   m=1024, ~72 MB at m=16K. Feasible for quarterly cadence.")
    print()
    print("7. Session 6 finding: single-Beneš mixing at m=4 has Friedrichs")
    print("   angle 60° exactly. If this holds at general m, T4 = 4/m²")
    print("   provides a clean statistical bound complementing T1's")
    print("   guessing bound.")
    print()
    print("=" * 70)
    print("OPERATING ENVELOPE (Regime 2, current deployment)")
    print("=" * 70)
    print()
    print("Under the current 'DP-off, both-blind + two-server enabled' regime,")
    print("the binding constraint is T2 = 1/(k + 1/(e^ε_cell − 1)).")
    print("For MAS default (k=25, ε_cell=1): T2 ≈ 1/25.58 = 0.039.")
    print()
    print("To tighten:")
    print("  (a) Increase k-anon threshold to 50 → T2 ≈ 0.020.")
    print("  (b) Increase padding budget (raise ε_cell) → T2 → 1/k as ε → ∞.")
    print("     At ε_cell → ∞: T2 → 1/25 = 0.040. Diminishing returns.")
    print("  (c) Both — k=50, ε_cell → ∞: T2 → 1/50 = 0.020.")
    print()
    print("PAPER RECOMMENDATION (revised for regime 2):")
    print("  - Crypto tier (T_crypto) is not the constraint. Both-blind +")
    print("    two-server protocol provides strong coalition-boundary")
    print("    guarantees at negligible Adv cost.")
    print("  - Set k-anon threshold to 50 (or higher) to bring T2 below 0.02.")
    print("  - No DP layer needed if the deployment target is 'k-anon per")
    print("    cell + one-shot Beneš'; T_crypto is astronomically small,")
    print("    T2 is the entire leakage floor.")
    print("  - Session 4/5 coalition improvements are the OPERATIONAL WIN:")
    print("    matcher (MAS) + any single bank collusion is now blocked from")
    print("    reversing the shuffle. Enable both-blind variant in production.")


if __name__ == "__main__":
    main()
